src/link.rs

```diff
@@ -27,101 +30,142 @@ impl Link {
         let p = path.as_ref().to_path_buf();
         let url = Self::load_url(p)?;
         let id = ResourceId::compute_bytes(url.as_str().as_bytes())?;
+        // Load user properties first
+        let user_prop = Self::load_user_data(&root, &id)?;
+        let mut description = user_prop.desc;
+
+        // Only load properties if the description is not set
+        if description.is_none() {
+            let bytes = load_raw_properties(root.as_ref(), id)?;
+            let graph_meta: OpenGraph = serde_json::from_slice(&bytes)?;
+            description = graph_meta.description;
+        }
 
-        let bytes = load_meta_bytes::<PathBuf>(root.as_ref().to_owned(), id)?;
-        let meta: Metadata =
-            serde_json::from_slice(&bytes).map_err(|_| ArklibError::Parse)?;
-
-        Ok(Self { url, meta })
+        Ok(Self {
+            url,
+            prop: Properties {
+                title: user_prop.title,
+                desc: description,
+            },
+        })
     }
 
-    /// Write zipped file to path
-    pub async fn write_to_path<P: AsRef<Path>>(
-        &mut self,
+    pub async fn save<P: AsRef<Path>>(
+        &self,
         root: P,
-        path: P,
-        save_preview: bool,
+        with_preview: bool,
     ) -> Result<()> {
         let id = self.id()?;
-        store_meta::<Metadata, _>(root.as_ref(), id, &self.meta)?;
-
-        let mut link_file = File::create(path.as_ref().to_owned())?;
-        let file_data = self.url.as_str().as_bytes();
-        link_file.write(file_data)?;
-        if save_preview {
-            let preview_data = Link::get_preview(self.url.clone())
-                .await
-                .unwrap_or_default();
-            let image_data = preview_data
-                .fetch_image()
-                .await
-                .unwrap_or_default();
-            self.save_preview(root.as_ref(), image_data, id)
-                .await?;
+        let id_string = id.to_string();
+        let base_dir = root.as_ref().join(ARK_FOLDER);
+        let folder = base_dir
+            .join(LINK_STORAGE_FOLDER)
+            .join(&id_string);
+        let link_file = AtomicFile::new(&folder)?;
+        let tmp = link_file.make_temp()?;
+        (&tmp).write_all(self.url.as_str().as_bytes())?;
+        let current_link = link_file.load()?;
+        link_file.compare_and_swap(&current_link, tmp)?;
+
+        //User defined properties
+        let prop_folder = base_dir
+            .join(PROPERTIES_STORAGE_FOLDER)
+            .join(&id_string);
+        let prop_file = AtomicFile::new(prop_folder)?;
+        modify_json(&prop_file, |data: &mut Option<Properties>| {
+            let properties = self.prop.clone();
+            match data {
+                Some(data) => {
+                    // Hack currently overwrites
+                    *data = properties;
+                }
+                None => *data = Some(properties),
+            }
+        })?;
```

Comment:
It looks like we should use `fn store_properties` here.

We also might need to change `struct Properties` declaration.

Reply:
Is there reason to not always use Vec<String> ? It would easy the process of deserialisation and allow to define a uniq type of properties instead of having either String or Vec<String>